### PR TITLE
Record method counts while profiling query components

### PR DESCRIPTION
**Pull request by polyfractal**
Records the number of times each Lucene method is invoked.  These counts can be used to help judge the selectivity of individual query components in the context of the entire query.  

E.g. a query may not look selective when run by itself (matches most of the index), but when run in context of a full search request, is evaluated only rarely due to execution order.

It seemed like a relatively minor overhead to profiling, and could provide useful information in the future.

/cc @jpountz thoughts?

### CODE REVIEW

**Comment by polyfractal** [comment]
/cc @colings86 too, since you're doing profile stuff now as well.

Note: this needs to be updated because of #18370

**Comment by colings86** [comment]
I like the idea. Having number of invocations on each timing type can also help see bottlenecks a bit clearer. For example, if you have two parts of your query that take 2 seconds on the profiler, but one has 1,000,000 invocations and the other has 4 then it is a good indication that the place to start looking is the query that is called 4 times

File: docs/reference/search/profile.asciidoc [hunk 1]
```diff
@@ -53,72 +53,90 @@ This will yield the following result:
       "failed": 0
    },
    "hits": {
-      "total": 1,
-      "max_score": 1,
+      "total": 3,
+      "max_score": 1.078072,
```
Comment (colings86) [inline docs/reference/search/profile.asciidoc]: it seems like most of the changes in this file are just changes to the timings of the parts of the queries. Is this page auto-generated or something?

Comment (polyfractal) [inline docs/reference/search/profile.asciidoc]: I copy/pasted a response to get all the new elements (didn't want to make up the counts), but then the timings no longer matched and my OCD wouldn't allow mismatching timings all over the page :)

**Comment by colings86** [comment]
@polyfractal are you still planning on progressing with this? it would be good to get this in IMO

**Comment by polyfractal** [comment]
Em, yes, sorry.  Got lost in the wedding/travel frenzy.  Will update this today or tomorrow.

**Comment by polyfractal** [comment]
Rebased, and updated the responses for the aggregation profiles since they include counts now too.

@colings86 mind taking a quick look at this?

**Comment by colings86** [comment]
@polyfractal LGTM, think this change will be really useful

File: docs/reference/search/profile.asciidoc [hunk 2]
```diff
@@ -699,6 +756,10 @@ The meaning of the stats are as follows:
 
     This is not currently used and will always report `0`. Currently aggregation profiling only times the shard level parts of the aggregation execution. Timing of the reduce phase will be added later.
 
+`*_count`::
+    Records the number of invocations of the particular method.  For example, `"collect": 2,`
```
Comment (jpountz) [inline docs/reference/search/profile.asciidoc]: do you mean `"collect_count": 2,`?

**Comment by jpountz** [comment]
LGTM too

File: docs/reference/search/profile.asciidoc (hunk 2 above)
Comment (polyfractal) [inline docs/reference/search/profile.asciidoc]: Oops, I did.  Fixing!

**Pull request merged**